### PR TITLE
Add blog post about Flink's rescalable state

### PR DESCRIPTION
**Pull request by wints**
This post gives readers an intro to stateful stream processing as well as a detailed overview of rescalable state in Flink. 

@StefanRRichter @rmetzger @tzulitai

### CODE REVIEW

File: _posts/2017-07-04-flink-rescalable-state.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+In the following section, we’ll explain how we solved the problem of efficient, meaningful state reassignment in Flink. Each of Flink state’s two flavours, operator state and keyed state, requires a different approach to state assignment.
+
+## Reassigning Operator State When Rescaling
+
+First, we’ll discuss how state reassignment in rescaling works for operator state. A common real-world use-case of operator state in Flink is to maintain the current offsets for Kafka partitions in Kafka sources. Each Kafka source instance would maintain `<PartitionID, Offset>` pairs–one pair for each Kafka partition that the source is reading–as operator state. How would we redistribute this operator state in case of rescaling? Ideally, we would like to reassign all `<PartitionID, Offset>` pairs from the checkpoint in round robin across all parallel operator instances after the rescaling.
+
+As a user, we are aware of the “meaning” of Kafka partition offsets, and we know that we can treat them as independent, redistributable units of state. The problem remains how we can we share this domain-specific knowledge with Flink.
```
Comment (tzulitai) [inline _posts/2017-07-04-flink-rescalable-state.md]: "The problem remains how we can we share this domain-specific knowledge with Flink."
maybe:
"The problem of how we can share this domain-specific knowledge with Flink remains."?

File: _posts/2017-07-04-flink-rescalable-state.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+We could again assign the state from previous `map_1` and `map_2` to the new `map_1` and `map_2`. But this would leave `map_3` with empty state. Depending on the type of state and concrete semantics of the job, this naive approach could lead to anything from inefficiency to incorrect results.
+
+In the following section, we’ll explain how we solved the problem of efficient, meaningful state reassignment in Flink. Each of Flink state’s two flavours, operator state and keyed state, requires a different approach to state assignment.
+
+## Reassigning Operator State When Rescaling
+
+First, we’ll discuss how state reassignment in rescaling works for operator state. A common real-world use-case of operator state in Flink is to maintain the current offsets for Kafka partitions in Kafka sources. Each Kafka source instance would maintain `<PartitionID, Offset>` pairs–one pair for each Kafka partition that the source is reading–as operator state. How would we redistribute this operator state in case of rescaling? Ideally, we would like to reassign all `<PartitionID, Offset>` pairs from the checkpoint in round robin across all parallel operator instances after the rescaling.
```
Comment (tzulitai) [inline _posts/2017-07-04-flink-rescalable-state.md]: "pairs-one pair for ..."

empty spaces should surround '-'.

File: _posts/2017-07-04-flink-rescalable-state.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+}
+```
+
+## Reassigning Keyed State When Rescaling
+The second flavour of state in Flink is keyed state. In contrast to operator state, keyed state is scoped by key, where the key is extracted from each stream event.
+
+To illustrate how keyed state differs from operator state, let’s use the following example. Assume we have a stream of events, where each event has the schema `{customer_id:int, value:int}`. We have already learned that we can use operator state to compute and emit the running sum of values for all customers.
+
+Now assume we want to slightly modify our goal and compute a running sum of values for each individual customer_id. This is a use case from keyed state, as one aggregated state must be maintained for each unique key in the stream.
```
Comment (tzulitai) [inline _posts/2017-07-04-flink-rescalable-state.md]: make "customer_id" be formatted as code, i.e. surround it with `'s => `customer_id`.

File: _posts/2017-07-04-flink-rescalable-state.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+While this automatically solves the problem of logically remapping the state to sub-tasks after rescaling, there is one more practical problem left to solve: how can we efficiently transfer the state to the subtasks’ local backends?
+
+When we’re not rescaling, each subtask can simply read the whole state as written to the checkpoint by a previous instance in one sequential read.
+
+When rescaling, however, this is no longer possible–the state for each subtask is now potentially scattered across the files written by all subtasks (think about what happens if you change the parallelism in `hash(key) mod parallelism`). We have illustrated this problem in **Figure 3A**. In this example, we show how keys are shuffled when rescaling from parallelism 3 to 4 for a key space of 0, 20, using identity as hash function to keep it easy to follow.
```
Comment (tzulitai) [inline _posts/2017-07-04-flink-rescalable-state.md]: "possible-the state for ..."

empty spaces should surround '-'.

**Comment by wints** [comment]
Thanks @tzulitai, great feedback! I've pushed a version with the edits you suggested.

**Comment by tzulitai** [comment]
Thanks! Merging this now ..

**Pull request closed without merging**